Asynchonous Entity Add When Generating Tree in the Wrong Place

### Stack trace

```
[14:10:19 ERROR]: Remove callback for map threw exception 
java.util.concurrent.CompletionException: net.minecraft.ReportedException: Exception generating new chunk
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155) ~[?:?]
        at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) ~[?:?]
        at net.minecraft.server.level.ChunkTaskPriorityQueueSorter.lambda$message$1(ChunkTaskPriorityQueueSorter.java:58) ~[app:?]
        at net.minecraft.util.thread.ProcessorMailbox.pollTask(ProcessorMailbox.java:87) ~[app:?]
        at net.minecraft.util.thread.ProcessorMailbox.pollUntil(ProcessorMailbox.java:141) ~[app:?]
        at net.minecraft.util.thread.ProcessorMailbox.run(ProcessorMailbox.java:96) ~[app:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) ~[?:?]
        at java.lang.Thread.run(Thread.java:831) ~[?:?]
Caused by: net.minecraft.ReportedException: Exception generating new chunk
        at net.minecraft.server.level.ChunkTracker.lambda$scheduleChunkGeneration$38(ChunkTracker.java:1340) ~[app:?]
        at com.mojang.datafixers.util.Either$Left.map(Either.java:38) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.server.level.ChunkTracker.lambda$scheduleChunkGeneration$40(ChunkTracker.java:1324) ~[app:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
        ... 8 more
Caused by: java.lang.IllegalStateException: Asynchronous entity add!
        at org.spigotmc.AsyncCatcher.catchOp(AsyncCatcher.java:16) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.server.level.ServerLevel.addEntity0(ServerLevel.java:1365) ~[app:?]
        at net.minecraft.server.level.ServerLevel.addFreshEntity(ServerLevel.java:1306) ~[app:?]
        at net.minecraft.world.level.block.Block.popResource(Block.java:404) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.world.level.block.Block.popResource(Block.java:370) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.world.level.block.Block.lambda$dropResources$4(Block.java:355) ~[patched_1.17.1.jar:git-Paper-408]
        at java.util.ArrayList.forEach(ArrayList.java:1511) ~[?:?]
        at net.minecraft.world.level.block.Block.dropResources(Block.java:354) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.server.level.WorldGenRegion.destroyBlock(WorldGenRegion.java:241) ~[app:?]
        at net.minecraft.world.level.LevelWriter.destroyBlock(LevelWriter.java:23) ~[app:?]
        at net.minecraft.world.level.LevelWriter.destroyBlock(LevelWriter.java:19) ~[app:?]
        at net.minecraft.world.level.levelgen.feature.HugeFungusFeature.placeStem(HugeFungusFeature.java:88) ~[app:?]
        at net.minecraft.world.level.levelgen.feature.HugeFungusFeature.place(HugeFungusFeature.java:60) ~[app:?]
        at org.bukkit.craftbukkit.v1_17_R1.CraftRegionAccessor.generateTree(CraftRegionAccessor.java:362) ~[patched_1.17.1.jar:git-Paper-408]
        at org.bukkit.craftbukkit.v1_17_R1.CraftRegionAccessor.generateTree(CraftRegionAccessor.java:276) ~[patched_1.17.1.jar:git-Paper-408]
        at org.bukkit.craftbukkit.v1_17_R1.generator.CraftLimitedRegion.generateTree(CraftLimitedRegion.java:181) ~[patched_1.17.1.jar:git-Paper-408]
        at io.github.addoncommunity.galactifun.base.universe.saturn.Titan$1.populate(Titan.java:162) ~[Galactifun - DEV 58 (git c9b57).jar:?]
        at net.minecraft.world.level.chunk.ChunkGenerator.addDecorations(ChunkGenerator.java:274) ~[patched_1.17.1.jar:git-Paper-408]
        at org.bukkit.craftbukkit.v1_17_R1.generator.CustomChunkGenerator.applyBiomeDecoration(CustomChunkGenerator.java:272) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.world.level.chunk.ChunkStatus.lambda$static$10(ChunkStatus.java:93) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.world.level.chunk.ChunkStatus.generate(ChunkStatus.java:212) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.server.level.ChunkTracker.lambda$scheduleChunkGeneration$38(ChunkTracker.java:1326) ~[app:?]
        at com.mojang.datafixers.util.Either$Left.map(Either.java:38) ~[patched_1.17.1.jar:git-Paper-408]
        at net.minecraft.server.level.ChunkTracker.lambda$scheduleChunkGeneration$40(ChunkTracker.java:1324) ~[app:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
        ... 8 more
```


### Plugin and Datapack List

Slimefun
Galactifun
I was not given any other info in the original report

### Actions to reproduce (if known)

_No response_

### Paper version

Paper 408 1.17.1

### Other

I am not the one who had this, see Slimefun-Addon-Community/Galactifun#116

--- a/patches/server/0945-Fix-async-entity-add-due-to-fungus-trees.patch
+++ b/patches/server/0945-Fix-async-entity-add-due-to-fungus-trees.patch
@@ -1,0 +1,35 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Jake Potrebic <jake.m.potrebic@gmail.com>
+Date: Fri, 18 Mar 2022 21:30:00 -0700
+Subject: [PATCH] Fix async entity add due to fungus trees
+
+
+diff --git a/src/main/java/net/minecraft/server/level/WorldGenRegion.java b/src/main/java/net/minecraft/server/level/WorldGenRegion.java
+index 80c1e0e47818486a68e0114b063395290365346b..0d3587e2d454fb0994ecff930b5e496a3999746a 100644
+--- a/src/main/java/net/minecraft/server/level/WorldGenRegion.java
++++ b/src/main/java/net/minecraft/server/level/WorldGenRegion.java
+@@ -227,6 +227,7 @@ public class WorldGenRegion implements WorldGenLevel {
+         if (iblockdata.isAir()) {
+             return false;
+         } else {
++            if (drop) LOGGER.warn("Potential async entity add during worldgen", new Throwable()); // Paper - log when this happens
+             if (false) { // CraftBukkit - SPIGOT-6833: Do not drop during world generation
+                 BlockEntity tileentity = iblockdata.hasBlockEntity() ? this.getBlockEntity(pos) : null;
+ 
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java b/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
+index 62bca85da6c5d9877e21fecb702370506ddf671c..c73468450bfe23e6f3a99d15c3e39a2a1ddf0c77 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftRegionAccessor.java
+@@ -372,10 +372,10 @@ public abstract class CraftRegionAccessor implements RegionAccessor {
+                 ((ChorusFlowerBlock) Blocks.CHORUS_FLOWER).generatePlant(access, pos, random, 8);
+                 return true;
+             case CRIMSON_FUNGUS:
+-                gen = TreeFeatures.CRIMSON_FUNGUS_PLANTED;
++                gen = this.isNormalWorld() ? TreeFeatures.CRIMSON_FUNGUS_PLANTED : TreeFeatures.CRIMSON_FUNGUS; // Paper - if world gen, don't use planted version
+                 break;
+             case WARPED_FUNGUS:
+-                gen = TreeFeatures.WARPED_FUNGUS_PLANTED;
++                gen = this.isNormalWorld() ? TreeFeatures.WARPED_FUNGUS_PLANTED : TreeFeatures.WARPED_FUNGUS; // Paper - if world gen, don't use planted version
+                 break;
+             case AZALEA:
+                 gen = TreeFeatures.AZALEA_TREE;

